### PR TITLE
consensus/babe: block inherents

### LINKED ISSUE #500: Get inherents data and apply to new block
In build block, we need to call `inherent_extrinsics` to get the inherents data(queue) to include in the block, then call `apply_extrinsic` for each inherent.

### PR DESCRIPTION
**Pull request by noot**
## Changes
- implement InherentsData struct as a mapping
- implement adding new fields and Encode for InherentsData
- construct and apply InherentsData in build-block

## Tests:
```
 go test ./consensus/...
```

### Issues:
closes #500

### CODE REVIEW

**Comment by noot** [comment]
@ryanchristo that was my initial thought but @ansermino says it is confusing. maybe we can have a discussion about this? I initially set my other branch to be the base so that the changes were clearer.

File: consensus/babe/types.go [hunk 1]
```diff
@@ -42,3 +48,68 @@ type AuthorityData struct {
...
+		}
+		_, err = buffer.Write(v)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return buffer.Bytes(), nil
+}
+
+type Slot struct {
```
Comment (ryanchristo) [inline consensus/babe/types.go]: Should this be at the top of the file? Also, a comment would be nice explaining `Slot`.

**Comment by ansermino** [comment]
This PR is not ready for review as it is based off code that is yet to be merged. Please use draft PRs in the future, or wait for the code to be merged.

**Comment by noot** [comment]
@ansermino ok will do

File: consensus/babe/types.go (hunk 1 above)
Comment (noot) [reply consensus/babe/types.go]: I've moved it and added a comment

File: consensus/babe/types.go [hunk 2]
```diff
@@ -51,14 +57,68 @@ type BabeHeader struct {
...
+		if err != nil {
+			return nil, err
+		}
+		_, err = buffer.Write(v)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return buffer.Bytes(), nil
+}
```
Comment (ansermino) [inline consensus/babe/types.go]: Please move to `inherents.go` or similar

File: consensus/babe/babe.go [hunk 3]
```diff
@@ -226,3 +230,30 @@ func (b *Session) buildBlock(parent *types.BlockHeaderWithHash, slot Slot) (*typ
 	block.Header.Number.Add(parent.Number, big.NewInt(1))
 	return block, nil
 }
+
+func (b *Session) buildBlockInherents(slot Slot) error {
```
Comment (ansermino) [inline consensus/babe/babe.go]: Please add a comment for this

File: consensus/babe/runtime.go [hunk 4]
```diff
@@ -39,3 +43,74 @@ func (b *Session) configurationFromRuntime() error {
 
 	return err
 }
+
+// calls runtime API function Core_initialize_block
+func (b *Session) initializeBlock(blockHeader []byte) error {
+	// TODO: use allocator to store block header
```
[the remark this answers is not in the document]
Comment (noot) [reply consensus/babe/runtime.go]: it's been fixed in my next PR

File: consensus/babe/babe.go (hunk 3 above)
Comment (noot) [reply consensus/babe/babe.go]: done

File: consensus/babe/types.go (hunk 2 above)
Comment (noot) [reply consensus/babe/types.go]: done

File: consensus/babe/runtime.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (noot) [reply consensus/babe/runtime.go]: https://github.com/ChainSafe/gossamer/pull/507

**Pull request merged**